connector/spanmetricsconnector/connector.go

```diff
@@ -0,0 +1,544 @@
+func durationToMillis(d time.Duration) float64 {
+	return float64(d.Nanoseconds()) / float64(time.Millisecond.Nanoseconds())
+}
+
+func mapDurationsToMillis(vs []time.Duration) []float64 {
+	vsm := make([]float64, len(vs))
+	for i, v := range vs {
+		vsm[i] = durationToMillis(v)
+	}
+	return vsm
+}
+
+// validateDimensions checks duplicates for reserved dimensions and additional dimensions. Considering
+// the usage of Prometheus related exporters, we also validate the dimensions after sanitization.
+func validateDimensions(dimensions []Dimension, skipSanitizeLabel bool) error {
+	labelNames := make(map[string]struct{})
+	for _, key := range []string{serviceNameKey, spanKindKey, statusCodeKey} {
+		labelNames[key] = struct{}{}
+		labelNames[sanitize(key, skipSanitizeLabel)] = struct{}{}
+	}
+	labelNames[operationKey] = struct{}{}
+
+	for _, key := range dimensions {
+		if _, ok := labelNames[key.Name]; ok {
+			return fmt.Errorf("duplicate dimension name %s", key.Name)
+		}
+		labelNames[key.Name] = struct{}{}
+
+		sanitizedName := sanitize(key.Name, skipSanitizeLabel)
+		if sanitizedName == key.Name {
+			continue
+		}
+		if _, ok := labelNames[sanitizedName]; ok {
+			return fmt.Errorf("duplicate dimension name %s after sanitization", sanitizedName)
+		}
+		labelNames[sanitizedName] = struct{}{}
+	}
+
+	return nil
+}
+
+// Start implements the component.Component interface.
+func (p *connectorImp) Start(ctx context.Context, host component.Host) error {
+	p.logger.Info("Starting spanmetricsprocessor")
+
+	p.started = true
+	go func() {
+		for {
+			select {
+			case <-p.done:
+				return
+			case <-p.ticker.C:
+				p.exportMetrics(ctx)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Shutdown implements the component.Component interface.
+func (p *connectorImp) Shutdown(context.Context) error {
+	p.shutdownOnce.Do(func() {
+		if p.tracesConsumer == nil {
+			p.logger.Info("Shutting down spanmetricsconnector")
+		} else {
+			p.logger.Info("Shutting down spanmetricsprocessor")
+		}
+		if p.started {
+			p.logger.Info("Stopping ticker")
+			p.ticker.Stop()
+			p.done <- struct{}{}
+			p.started = false
+		}
+	})
+	return nil
+}
+
+// Capabilities implements the consumer interface.
+func (p *connectorImp) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+// ConsumeTraces implements the consumer.Traces interface.
+// It aggregates the trace data to generate metrics, forwarding these metrics to the discovered metrics exporter.
+// The original input trace data will be forwarded to the next consumer, unmodified.
+func (p *connectorImp) ConsumeTraces(ctx context.Context, traces ptrace.Traces) error {
+	p.lock.Lock()
+	p.aggregateMetrics(traces)
+	p.lock.Unlock()
+
+	// true when p is a connector
+	if p.tracesConsumer == nil {
+		return nil
+	}
+
+	// Forward trace data unmodified and propagate trace pipeline errors, if any.
+	return p.tracesConsumer.ConsumeTraces(ctx, traces)
```

Comment:
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18535/files#diff-78f9e7446e0e4fb474786b4b2c262486887bf27479f1f8bc598376d6c3c29b72R283-R289
Since we are not supporting `traces to traces` pipeline, we should return nil.

Reply:
done